### PR TITLE
Make non-divisible splits not change extents used in indexing

### LINKED ISSUE #1074: Invalid striding
This is related to #1016. I thought that was fixed by #1018, but that isn't actually sufficient.

```
  auto tv0 = makeSymbolicTensor(2);
  fusion.addInput(tv0);
  auto tv1 = add(tv0, new Double(1));
  auto tv2 = add(tv1, new Double(1));
  fusion.addOutput(tv2);

  tv2->split(0, 2);
  tv2->split(-1, 4);
  tv2->reorder({{1, 2}, {2, 1}});
  tv0->computeAt(tv2, 2);

  tv2->split(-1, 3);

  // To make the sanitizer catch the invalid accesses. Not necessary
  // to expose the bug.
  tv1->setMemoryType(MemoryType::Shared);
```

This generates:

```
Inputs:
  T0_g[ iS14{( ceilDiv(i1, 2) )}, iS16{( ceilDiv(i3, 4) )}, iS15{2}, iS17{4} ], float
Outputs:
  T2_g[ iS6{( ceilDiv(i1, 2) )}, iS8{( ceilDiv(i3, 4) )}, iS7{2}, iS18{( ceilDiv(4, 3) )}, iS19{3} ] produce_pos( 2), float

%kernel_math {
T1_s[ iS10{( ceilDiv(i1, 2) )}, iS12{( ceilDiv(i3, 4) )}, iS11{2}, iS13{4} ] ca_pos( 2 )
   = T0_g[ iS14{( ceilDiv(i1, 2) )}, iS16{( ceilDiv(i3, 4) )}, iS15{2}, iS17{4} ]
   + double(1);
T2_g[ iS6{( ceilDiv(i1, 2) )}, iS8{( ceilDiv(i3, 4) )}, iS7{2}, iS18{( ceilDiv(4, 3) )}, iS19{3} ] produce_pos( 2)
   = T1_s[ iS10{( ceilDiv(i1, 2) )}, iS12{( ceilDiv(i3, 4) )}, iS11{2}, iS13{4} ] ca_pos( 2 )
   + double(1);
}

__global__ void CUDAGeneratedKernel(Tensor<float, 2> T0, Tensor<float, 2> T2) {
  NVFUSER_DEFINE_MAGIC_ZERO
  #pragma unroll 1
  for(nvfuser_index_t ki78 = 0; ki78 < (ceilDiv(T0.size[0], 2)); ++ki78) {
    #pragma unroll 1
    for(nvfuser_index_t ki80 = 0; ki80 < (ceilDiv(T0.size[1], 4)); ++ki80) {
      __shared__ float T1[(2 * 4)];
      #pragma unroll
      for(nvfuser_index_t ki91 = 0; ki91 < 2; ++ki91) {
        #pragma unroll
        for(nvfuser_index_t ki93 = 0; ki93 < 4; ++ki93) {
          if (((((ki78 * 2) + ki91) < T0.size[0]) && (((ki80 * 4) + (ki93 + nvfuser_zero)) < T0.size[1]))) {
            T1[(ki91 * 4) + ki93]
              = T0[(((ki78 * 2) + ki91) * T0.stride[0]) + (((ki80 * 4) + (ki93 + nvfuser_zero)) * T0.stride[1])]
              + (float) 1;
          }
        }
      }
      NVFUSER_UPDATE_MAGIC_ZERO
      __barrier_sync(0);
      #pragma unroll
      for(nvfuser_index_t ki82 = 0; ki82 < 2; ++ki82) {
        #pragma unroll
        for(nvfuser_index_t ki84 = 0; ki84 < (ceilDiv(4, 3)); ++ki84) {
          #pragma unroll
          for(nvfuser_index_t ki86 = 0; ki86 < 3; ++ki86) {
            if (((((ki78 * 2) + ki82) < T0.size[0]) && (((ki80 * ((ceilDiv(4, 3)) * 3)) + ((ki84 * 3) + (ki86 + nvfuser_zero))) < T0.size[1]))) {
              T2[(((ki78 * 2) + ki82) * (1 * T0.size[1])) + (((ki80 * ((ceilDiv(4, 3)) * 3)) + ((ki84 * 3) + (ki86 + nvfuser_zero))) * 1)]
                = T1[(ki82 * ((ceilDiv(4, 3)) * 3)) + ((ki84 * 3) + ki86)]
                + (float) 1;
            }
          }
        }
      }
      NVFUSER_UPDATE_MAGIC_ZERO
      __barrier_sync(0);
    }
  }
}
```

The problem is the read access to `T1[(ki82 * ((ceilDiv(4, 3)) * 3)) + ((ki84 * 3) + ki86)]`. The extent of the second axis is computed as `(ceilDiv(4, 3)) * 3)`, however, it should be just `4`. 
 
Here's a relevant part of the code: https://github.com/csarofeen/pytorch/blob/20_12_3_devel/torch/csrc/jit/codegen/cuda/index_compute.cpp#L503-L509

It seems updating the extent should be only done when indexing as a consumer, but then we'd need to make sure the proper extent is already available in the extent map when indexing as a producer.

### PR DESCRIPTION
**Pull request by naoyam**
This is a second attempt to fix an indexing problem that is first reported in PR #1243.  See that PR as well for the context of this problem.

Also, the repro of issue #1074 is now working (`FusionNonDivisibleSplit2`).

This PR addresses an index linearlization problem that was exposed by view and strided gather operations, both of which are based on rFactor. The root cause of the problem is non-divisible splits. When indexing with such splits, the domain that is split gets a larger extent due to the `ceilDiv` of the outer domain. This is propagated back from leaf to parent domains in `IndexCompute::handle(Split*)`. It turns out that there are consistency problems when to propagate. The most sensible way is likely to completely stop doing so, but it also means intermediate domains have to be predicated if split by a non-divisible factor.

Also, predicating such domains doesn't work if vectorization is involved. If an output domain of a split is vectorized, the split must be divisible. Divisibility is validated at the kernel launch time.

The main logic is in `NonDivisibleSplitInfo`. It scans a fusion to locate any `Split` expression that must be either predicated or validated. If an `IterDomain` is run-time validated, it's not predicated as it's unnecessary (`NonDivisibleSplitInfo::removeRedundancy()`). 

`getReferenceRootPredicates` is extended to generate predicates for those domains that may be split with a non-divisible factor. Currently, it finds root or merged contiguous domains to predicate. I extended that to include non-divisible split domains (`getNonDivisibleDomainsToPredicate`). While `index_compute.cpp` may appear to have many changes, most of them don't change the existing logic. For example, the change in `getStartAndStopOffsets` is not to do adjustments for shift and gather operations when predicating non-divisible split domains.

All of the C++ tests are passing, but `test_native_layer_norm_half` is failing due to a runtime validation of a non-divisible domain with vectorization. The domain to split is:

```
i515 = T0.size[1] * T0.size[2]
~ i518 = T0.size[0] * i515
~ i521 = ceilDiv(i518, 4)
i521
```

This `i521` domain is split by `2` and the inner domain of size 2 is vectorized. The test fails with `T0` being `[17, 10, 10]` since `17 * 10 * 10 / 4` isn't divisible by `2`. Without the runtime check, an alignment error occurs. Since @csarofeen mentioned he's working on changing scheduling algorithms for vectorization, I've done nothing on the failure yet.

Closes #1074

### CODE REVIEW

File: test/cpp/jit/test_gpu.cpp [hunk 1]
```diff
@@ -18822,6 +18822,318 @@ TEST(NVFuserTest, FusionRfactorPredication1_CUDA) {
       &fusion, cg_outputs, {at_t0, at_t3}, {at_t2, at_t4}, __LINE__, __FILE__);
 }
 
+TEST(NVFuserTest, FusionRfactorPredication2_CUDA) {
```
Comment (naoyam) [inline test/cpp/jit/test_gpu.cpp:18826]: This was a test created for #1250 but removed as it was failing due to the extent problem.

File: torch/csrc/jit/codegen/cuda/executor_utils.cpp [hunk 2]
```diff
@@ -405,6 +405,30 @@ bool canVectorize(
...
+// found, Vectorize is illegal.
+void validateVectorizedSplits(
+    kir::Kernel* kernel,
+    kir::ExpressionEvaluator& expr_eval) {
+  for (const auto& extent_factor : kernel->summary().splits_to_validate) {
+    auto input_extent = expr_eval.evaluate(extent_factor.first);
+    auto split_factor = expr_eval.evaluate(extent_factor.second);
+    TORCH_INTERNAL_ASSERT(
+        input_extent.has_value(),
+        "Extent not possible to evaluate: ",
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/executor_utils.cpp]: Can you mention in these assert messages this is from vectorizedSplit validation. A little more context in these messages will be helpful.

File: torch/csrc/jit/codegen/cuda/non_divisible_split.h [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+  //! run time
+  void removeRedundancy();
+
+ private:
+  //! Split expressions whose input domain must be predicated
+  std::unordered_map<TensorView*, std::vector<Split*>> splits_to_predicate_;
+  //! Split expressions whose divisbility must be validated at run time
+  std::unordered_set<Split*> splits_to_validate_;
+
+  //! Temporary used for analyzing each tensor
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/non_divisible_split.h]: Nit: Temporarily

File: torch/csrc/jit/codegen/cuda/non_divisible_split.h [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+  Val* getMaybeNonDivisibleExtent(Split* split) const;
+
+  //! Remove redundant predicates as divisibility may be validated at
+  //! run time
+  void removeRedundancy();
+
+ private:
+  //! Split expressions whose input domain must be predicated
+  std::unordered_map<TensorView*, std::vector<Split*>> splits_to_predicate_;
+  //! Split expressions whose divisbility must be validated at run time
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/non_divisible_split.h]: Nit: divisibility

File: torch/csrc/jit/codegen/cuda/lower2device.h [hunk 5]
```diff
@@ -111,6 +112,14 @@ class TORCH_CUDA_CU_API GpuLower {
     return partial_split_map_;
   }
 
+  auto& nonDivisibleSplitInfo() {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower2device.h:115]: Should non-const references in this header be private? Maybe we should check this as a follow up.

File: torch/csrc/jit/codegen/cuda/index_compute.cpp [hunk 6]
```diff
@@ -2941,13 +3010,14 @@ std::pair<std::vector<RootPredicateInfo>, ReferenceTensor> Index::
     }
 
     // Find a corresponding consumer root id if exists. Used to
-    // supprot shift. If contig_id is not root, nothing is required to
-    // do for shift as shift-related domains are excluded from
-    // contig domains.
+    // supprot shift. If contig_id a merged non-root domain, nothing
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/index_compute.cpp]: Nit: **support** shift. If **a** contig_id **is** a merged non-root domain, nothing

**Review by csarofeen — APPROVED** [review]
Should we clean up the predication logic to always use an explicit list of which iteration domains need to be predicated? I understand why this was only done for non-divisible-splits, but it seems to me we could have a pass that would specify all the domains that need to be predicated, including if root domains need predication, or just the contiguous merge domains. Maybe it would be good to try and unify this logic a bit more.

Looks really good overall, approving!

File: torch/csrc/jit/codegen/cuda/lower2device.h (hunk 5 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/lower2device.h:115]: As long as `GpuLower` is passed as a const object after lowering, I think it should be fine, but yes, this class is growing and may need some refactoring.

**Comment by naoyam** [comment]
> Should we clean up the predication logic to always use an explicit list of which iteration domains need to be predicated? I understand why this was only done for non-divisible-splits, but it seems to me we could have a pass that would specify all the domains that need to be predicated, including if root domains need predication, or just the contiguous merge domains. Maybe it would be good to try and unify this logic a bit more.

Agreed. We have contig IDs, non-divisible split IDs, and non-exact parallelized IDs. The last one is added in `getInlinePredicate`, but there's no reason that all three can't be done uniformly. I'll revisit later.

**Pull request merged**